### ISSUE TITLE
nl.adaptivity.xmlutil.serialization.UnknownXmlFieldException: Could not find a field for name

### ISSUE THREAD

**Issue by SchweinchenFuntik** [pdvrieze/xmlutil#107]
```kotlin
@Serializable
@XmlSerialName("repeat", Namespaces.view, Namespaces.prefixView)
data class Repeat(
    val value: String = "",
    val elements: List<Element> = emptyList()
) : Element

@Serializable
@XmlSerialName("option", Namespaces.view, Namespaces.prefixView)
data class Option(@Required val value: String) : Element

@Serializable
@XmlSerialName("select", Namespaces.view, Namespaces.prefixView)
data class Select(
    val values: String = "",
    val elements: List<Node> = emptyList()
) : Element {

    @Serializable
    sealed interface Node // OriginalOption == Option up code
    @JvmInline @Serializable value class Repeat(val repeat: OriginalRepeat) : Node
    @JvmInline @Serializable value class Option(val repeat: OriginalOption) : Node
}

fun main() {
    val xml = XML {
        indent = 4
        autoPolymorphic = true
        encodeDefault = XmlSerializationPolicy.XmlEncodeDefault.NEVER
    }
    
    val document = Document(
        head = Head(),
        body = Body(
            elements = listOf(
                Select(variable = "foo", elements = listOf(Select.Option(Option("test")))),
            )
        ),
    )
    
    val strXml = xml.encodeToString(document) // ok
    println(strXml)
    val decodeDoc = xml.decodeFromString<Document>(strXml) // error
    println(decodeDoc)
}
```

stascktrace
```
Exception in thread "main" nl.adaptivity.xmlutil.serialization.UnknownXmlFieldException: Could not find a field for name (org.xml.view.Action) {http://www.webui-ap.com/schemata/view/1.0}action/{http://www.webui-ap.com/schemata/view/1.0}meta (Element)
  candidates: title, type, {http://www.webui-ap.com/schemata/view/1.0}Event, when, name, on, value, result, close-on-escape, close-on-lost-focus, show-close-button, darken-background, darken-background-content-highlight, new-tab, dismissible, modal-style, top, bottom, right, left, target, show-title, window-name, command-line-args, timeout, delay, has-priority, {http://www.webui-ap.com/schemata/view/1.0}Action, {http://www.webui-ap.com/schemata/view/1.0}Meta, {http://www.webui-ap.com/schemata/view/1.0}Parameter at position Line number = 14
Column number = 49
System Id = null
Public Id = null
Location Uri= null
CharacterOffset = 560

	at nl.adaptivity.xmlutil.serialization.XmlConfig.DEFAULT_UNKNOWN_CHILD_HANDLER$lambda-2(XmlConfig.kt:365)
	at nl.adaptivity.xmlutil.serialization.DefaultXmlSerializationPolicy.handleUnknownContentRecovering(XmlSerializationPolicy.kt:431)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.indexOf(XMLDecoder.kt:588)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeElementIndex(XMLDecoder.kt:693)
	at org.xml.view.Action$$serializer.deserialize(Action.kt:15)
	at org.xml.view.Action$$serializer.deserialize(Action.kt:15)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeSerializableElement(XMLDecoder.kt:461)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$PolymorphicDecoder.decodeSerializableElement(XMLDecoder.kt:1203)
	at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:537)
	at kotlinx.serialization.internal.AbstractPolymorphicSerializer.deserialize(AbstractPolymorphicSerializer.kt:57)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$AnonymousListDecoder.decodeSerializableElement(XMLDecoder.kt:905)
	at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:537)
	at kotlinx.serialization.internal.CollectionLikeSerializer.readElement(CollectionSerializers.kt:80)
	at kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:51)
	at kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:36)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeSerializableElement(XMLDecoder.kt:459)
	at org.xml.view.Body$$serializer.deserialize(body.kt:8)
	at org.xml.view.Body$$serializer.deserialize(body.kt:8)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeSerializableElement(XMLDecoder.kt:461)
	at org.xml.view.Document$$serializer.deserialize(document.kt:16)
	at org.xml.view.Document$$serializer.deserialize(document.kt:16)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$XmlDecoder.decodeSerializableValue(XMLDecoder.kt:195)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromReader(XML.kt:412)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromReader$default(XML.kt:383)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromString(XML.kt:349)
	at org.funtik.mscada.MainKt.main(Main.kt:69)
	at org.funtik.mscada.MainKt.main(Main.kt)

Process finished with exit code 1

```

**Comment by pdvrieze** [comment]
There is a candidate `{http://www.webui-ap.com/schemata/view/1.0}Meta` but the tag found is `{http://www.webui-ap.com/schemata/view/1.0}meta` - it appears that there is a capitalisation error/you didn't declare a `serialName`/`XmlSerialName` annotation on the type. (I don't have the code, so can't quite confirm it)

The same with Parameter and Action. Note that there is some magic logic happening to determine tag names (and namespaces) when no annotation is given. The logic can be a bit complex on how this works (for the default policy) although using `XmlSerialName` will make this better.

**Comment by SchweinchenFuntik** [comment]
the scheme is the same as in #106 (I give the full code)
```kotlin
package issue.bug

import kotlinx.serialization.*
import nl.adaptivity.xmlutil.serialization.*
import issue.bug.Repeat as RealRepeat
import issue.bug.Option as RealOption

@Serializable
sealed interface Element // Any

fun main() {
    val xml = XML {
        indent = 4
        autoPolymorphic = true
        encodeDefault = XmlSerializationPolicy.XmlEncodeDefault.NEVER
    }

    issue107(xml)
}

fun issue107(xml: XML) {
    val document = Document(
        head = Head("ISSUE 107"),
        body = Body(
            elements = listOf(
                Select(variable = "foo", elements = listOf(Select.Option(RealOption("test")))),
            )
        ),
    )

    val strXml = xml.encodeToString(document) // ok
    println(strXml)
    val decodeDoc = xml.decodeFromString<Document>(strXml) // error
    println(decodeDoc)
}
// #106 
@Serializable
@XmlSerialName("document", Namespaces.view, Namespaces.prefixView)
data class Document(
    val head: Head,
    val body: Body,
    val nature: String? = null, // DEFAULT XMLNS
    @XmlOtherAttributes
    val other: Map<String, String> = emptyMap()
)
@Serializable
@XmlSerialName("head", Namespaces.view, Namespaces.prefixView)
data class Head(@XmlElement(true) val title: String? = null)


@Serializable
@XmlSerialName("body",  Namespaces.view,  Namespaces.prefixView)
data class Body(
    val elements: List<Element> = emptyList(),
    @XmlOtherAttributes
    val other: Map<String, String> = emptyMap()
)

@Serializable
@XmlSerialName("select", Namespaces.view, Namespaces.prefixView)
data class Select(
    val variable: String = "",
    val elements: List<Node> = emptyList()
) : Element {

    @Serializable
    sealed interface Node // typing scheme
    @JvmInline @Serializable value class Repeat(val repeat: RealRepeat) : Node
    @JvmInline @Serializable value class Option(val repeat: RealOption) : Node
}

@Serializable
@XmlSerialName("repeat", Namespaces.view, Namespaces.prefixView)
data class Repeat(
    val text: String = "",
    val elements: List<Element> = emptyList()
) : Element


@Serializable
@XmlSerialName("option", Namespaces.view, Namespaces.prefixView)
data class Option(@Required val value: String, val text: String = "", val selected: String = "") : Element


object Namespaces {
    const val view = "http://custom"
    const val prefixView = "d"
}
```

**output**
```
<d:document xmlns:d="http://custom">
    <d:head>
        <d:title>ISSUE 107</d:title>
    </d:head>
    <d:body>
        <d:select variable="foo">
            <d:option value="test"/>
        </d:select>
    </d:body>
</d:document>
Exception in thread "main" nl.adaptivity.xmlutil.serialization.UnknownXmlFieldException: Could not find a field for name (issue.bug.Select) {http://custom}select/{http://custom}option (Element)
  candidates: variable, {http://custom}Option, {http://custom}Repeat at position Line number = 7
Column number = 37
System Id = null
Public Id = null
Location Uri= null
CharacterOffset = 189

	at nl.adaptivity.xmlutil.serialization.XmlConfig.DEFAULT_UNKNOWN_CHILD_HANDLER$lambda-2(XmlConfig.kt:365)
	at nl.adaptivity.xmlutil.serialization.DefaultXmlSerializationPolicy.handleUnknownContentRecovering(XmlSerializationPolicy.kt:431)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.indexOf(XMLDecoder.kt:588)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeElementIndex(XMLDecoder.kt:693)
	at issue.bug.Select$$serializer.deserialize(main.kt:76)
	at issue.bug.Select$$serializer.deserialize(main.kt:76)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeSerializableElement(XMLDecoder.kt:461)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$PolymorphicDecoder.decodeSerializableElement(XMLDecoder.kt:1203)
	at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:537)
	at kotlinx.serialization.internal.AbstractPolymorphicSerializer.deserialize(AbstractPolymorphicSerializer.kt:57)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$AnonymousListDecoder.decodeSerializableElement(XMLDecoder.kt:905)
	at kotlinx.serialization.encoding.CompositeDecoder$DefaultImpls.decodeSerializableElement$default(Decoding.kt:537)
	at kotlinx.serialization.internal.CollectionLikeSerializer.readElement(CollectionSerializers.kt:80)
	at kotlinx.serialization.internal.AbstractCollectionSerializer.readElement$default(CollectionSerializers.kt:51)
	at kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:36)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeSerializableElement(XMLDecoder.kt:459)
	at issue.bug.Body$$serializer.deserialize(main.kt:68)
	at issue.bug.Body$$serializer.deserialize(main.kt:68)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$TagDecoder.decodeSerializableElement(XMLDecoder.kt:461)
	at issue.bug.Document$$serializer.deserialize(main.kt:54)
	at issue.bug.Document$$serializer.deserialize(main.kt:54)
	at nl.adaptivity.xmlutil.serialization.XmlDecoderBase$XmlDecoder.decodeSerializableValue(XMLDecoder.kt:195)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromReader(XML.kt:412)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromReader$default(XML.kt:383)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromString(XML.kt:349)
	at issue.bug.MainKt.issue107(main.kt:115)
	at issue.bug.MainKt.main(main.kt:19)
	at issue.bug.MainKt.main(main.kt)

Process finished with exit code 1

```

**Comment by SchweinchenFuntik** [comment]
if you set the flag `isCollectingNSAttributes = true`, then there will be an error as in #106

**Comment by pdvrieze** [comment]
I've managed to find (and resolve in dev) the issue. The problem is an error in the handling of inline/value classes when they are used polymorphically. The workaround is to put an @XmlSerialName annotation on the wrapping value classes.

As an aside you may want to consider having the outer repeat/option classes just implement node (that would avoid the need for the value classes entirely).

**Comment by pdvrieze** [comment]
The corrected handling is released in [0.85.0](https://github.com/pdvrieze/xmlutil/releases/tag/v0.85.0).
